Added an OrderedDictionary

TylerBrinkley opened the pull request:
No documentation yet.

Closes #2456

ahsonkhan commented on tests/Microsoft.Experimental.Collections.Tests/Microsoft/Collections/Extensions/OrderedDictionaryTests.cs:
```diff
@@ -0,0 +1,276 @@
+
+            // Tests OrderedDictionary Enumeration
+            var temp = 0;
+            foreach (var pair in test)
```
nit: avoid using var here (and for `var temp`) since it isn't clear from reading the right hand side what the type is.

TylerBrinkley replied:
:) It's hard to break old habits.

ahsonkhan commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/HashHelpers.cs:
```diff
@@ -0,0 +1,93 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
```
nit: remove extra new line

TylerBrinkley replied:
That's a direct copy/paste from coreclr.

ahsonkhan replied:
Should these remain in sync with https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Collections/HashHelpers.cs?

If so, can you please add a comment stating where you got this from.

ahsonkhan replied:
Also, regarding the nit, we should remove the newline from both places :)

Other files don't have the extra new lines, for example see: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Numerics/Hashing/HashHelpers.cs

ahsonkhan commented on tests/Microsoft.Experimental.Collections.Tests/Microsoft/Collections/Extensions/OrderedDictionaryTests.cs:
```diff
@@ -0,0 +1,276 @@
+    public class OrderedDictionaryTests
+    {
+        [Fact]
+        public void TestBasicOrderedDictionaryOperations()
```
nit: I would break these into separate API specific tests.

TylerBrinkley replied:
I agree. Is there anywhere within the .NET Core repos with guidance on how to write good unit tests, especially in regards to generic collections.

ahsonkhan commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,768 @@
+            {
+                throw new ArgumentNullException(nameof(array));
+            }
+            if (arrayIndex < 0 || arrayIndex > array.Length)
```
nit: re-write as a single if check
```C#
if ((uint)arrayIndex > (uint)array.Length)
```

ahsonkhan commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,768 @@
+            }
+
+            Entry[] entries = _entries;
+            for (int i = 0; i < count; ++i)
```
Is it guaranteed that `count` is within `entries.Length`?

TylerBrinkley replied:
Yes, the `Count` property is guaranteed to always be less than or equal to `_entries.Length`.

ahsonkhan replied:
Great. Debug.Assert?

danmosemsft replied:
A Debug.Assert would not add anything because it would throw on the next line.

ahsonkhan commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,768 @@
+            }
+        }
+
+        public sealed class KeyCollection : IList<TKey>, IReadOnlyList<TKey>
```
nit: split this out into a separate file

ahsonkhan commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,768 @@
+            }
+        }
+
+        public sealed class ValueCollection : IList<TValue>, IReadOnlyList<TValue>
```
same here: move to a separate file

ahsonkhan commented on tests/Microsoft.Experimental.Collections.Tests/Microsoft/Collections/Extensions/OrderedDictionaryTests.cs:
```diff
@@ -0,0 +1,276 @@
+        [Fact]
+        public void TestLargeOrderedDictionaryOperations()
+        {
+            const int count = 1000000;
```
nit: use digit separators for readability (`1_000_000`)

TylerBrinkley commented:
I've split out the tests into individual unit tests and have added support for many others.

Frassle commented:
Could we get interfaces to match, IOrderedDictionary and IReadOnlyOrderedDictionary? I've had code that could of benefited from IReadOnlyOrderedDictionary where Keys and Values are IList not IEnumerable.

danmosemsft commented:
This can close #2456, but probably we should have some issue tracking getting it through API and possibly moving it into CoreFX later.

danmosemsft commented:
@TylerBrinkley apologies for the delay.

@stephentoub @ahsonkhan @safern any further comments? Note, this is not waiting on API review - that's the purpose of creating the pack in this repo. However any review feedback is welcome including to the API. Note original discussion was in https://github.com/dotnet/corefx/issues/26634.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+        {
+            if (index < 0 || index >= Count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index), "Index was out of range. Must be non-negative and less than the size of the collection.");
```
Please use only strings from resx (although we're not localizing, all the other code does)

eg `SR.ArgumentOutOfRange_Index`

danmosemsft replied:
Same elsewhere

safern replied:
Also should we use a ThrowHelper for collections in this experimental package as well? To minimize IL size?

https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/ThrowHelper.cs

danmosemsft replied:
I don't think it's very important at this point.

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/IDictionaryDebugView.cs:
```diff
@@ -0,0 +1,82 @@
+
+        public DictionaryKeyCollectionDebugView(ICollection<TKey> collection)
+        {
+            if (collection == null)
```
this can be simplified to `_collection = collection ?? throw new ...`

TylerBrinkley replied:
That's a copy paste from Corefx.

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/IDictionaryDebugView.cs:
```diff
@@ -0,0 +1,82 @@
+
+        public DictionaryValueCollectionDebugView(ICollection<TValue> collection)
+        {
+            if (collection == null)
```
Ditto

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.KeyCollection.cs:
```diff
@@ -0,0 +1,137 @@
+
+            public KeyCollection(OrderedDictionary<TKey, TValue> orderedDictionary)
+            {
+                if (orderedDictionary == null)
```
Simplify with ??

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.KeyCollection.cs:
```diff
@@ -0,0 +1,137 @@
+                {
+                    throw new ArgumentNullException(nameof(array));
+                }
+                if (arrayIndex < 0 || arrayIndex > array.Length)
```
Use (uint) casting as @ahsonkhan previously suggested

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.ValueCollection.cs:
```diff
@@ -0,0 +1,145 @@
+
+            public ValueCollection(OrderedDictionary<TKey, TValue> orderedDictionary)
+            {
+                if (orderedDictionary == null)
```
Simplify with ??

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.ValueCollection.cs:
```diff
@@ -0,0 +1,145 @@
+                {
+                    throw new ArgumentNullException(nameof(array));
+                }
+                if (arrayIndex < 0 || arrayIndex > array.Length)
```
use uint casting

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+
+        public TValue GetAt(int index, out TKey key)
+        {
+            if (index < 0 || index >= Count)
```
(uint) ?

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+        {
+            int i = IndexOf(key, out int hashCode);
+            Entry[] entries = _entries;
+            ++_version;
```
You changed all postfix to prefix - personal style? I don't think it matters, although it increases the diff with Dictionary.

TylerBrinkley replied:
Just personal style from learning c++ in my college days. I can change it if need be.

safern replied:
I think it makes sense to be consistent with Dictionary, specially if we're not fixing the other nits because they came from corefx itself. Like the boundary checks by using uint.

safern replied:
Ahh github showed me an old diff. I saw you indeed fixed them :smile: thanks!

danmosemsft replied:
Our general pattern (in corefx anyway) seems to be to avoid both pre and post fix within an expression (eg `foo = bar++` or `foo = ++bar`) therefore the difference does not matter and we generally pick postfix as a convention.

Of course if the code is just clearer then exceptions are fine.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+                Array.Clear(_entries, 0, Count);
+                Count = 0;
+            }
+            ++_version;
```
I think `++_version` could be removed here if it's OK that enumeration just terminates. This was already done for Dictionary.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+            }
+
+            _entries[index].Value = value;
+            ++_version;
```
We could choose to avoid `++_version` for `SetAt()` and for `SetValue()` which ended up updating an existing value, because that will not disrupt enumeration. I am not sure why Dictionary increments it in that case.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+            }
+            Array.Copy(entries, actualIndex, entries, actualIndex + 1, count - actualIndex);
+
+            Debug.Assert(_buckets.Length > 0);
```
This assert is not useful because it would throw on the next line.

danmosemsft replied:
Same below

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+            }
+
+            // Remove the entry from the bucket
+            UpdateBucket(index, true);
```
Please use named parameter style as it saves looking at the callee. Eg `UpdateBucket(index, setToNext:true);`
Same elsewhere.
@stephentoub should that be in the [style guide](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md)?

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+        private readonly IEqualityComparer<TKey> _comparer;
+        private KeyCollection _keys;
+        private ValueCollection _values;
+        
```
I assume you removed `_freeList` because of the complexity? That's fine.

danmosemsft replied:
Oh, because you're coalescing entries 😄

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+            return true;
+        }
+
+        private void UpdateBucket(int i, bool setToNext, int incrementAmount = 0)
```
This method is hard to read because it is overloaded for two purposes: removing an entry from the bucket, and re-numbering the index in the bucket. I think it would be clearer to have two methods. It might even be faster (fewer branches)

danmosemsft commented on tests/Microsoft.Experimental.Collections.Tests/Microsoft/Collections/Extensions/OrderedDictionaryTests.cs:
```diff
@@ -0,0 +1,966 @@
+        }
+
+        [Fact]
+        public void TestWithStringComparerIgnoreCase()
```
Do you have tests for hash collision? OrdinalIgnoreCase could be used for a collision of 2 entries (although this test doesn't do that) but it would probably be worth having a custom comparer to test behavior when > 2 entries collide. Eg., repeat some of the general tests using a comparer that returns the same hashcode for everything.

danmosemsft replied:
I did this for DictionarySlim, they can share: https://github.com/dotnet/corefxlab/commit/da5f15991f7a9778f65c948bd54a646adebc10a1

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+            public int Next; // the index of the next item in the same bucket
+        }
+
+        private int[] _buckets = Array.Empty<int>();
```
Any particular reason you chose to initialize by default to capacity 0, compared with `Dictionary<TKey, TValue>` which looks like it will never have arrays smaller than 3? 

Although I wonder whether Dictionary should do the same. It makes empty ones cheaper.

danmosemsft replied:
Correction, I see Dictionary starts at 0 with null arrays, you just exchange a test for null array against a test for empty array. Maybe it makes no difference.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+            if (buckets.Length > 0)
+            {
+                int bucket = hashCode % buckets.Length;
+                index = buckets[bucket] - 1;
```
Consider adding GetBucketIndex and GetEntryIndex (eg like [this](https://github.com/dotnet/corefxlab/pull/2458/files#diff-6f44e9ce636799efd5c8b26382f2248bR39) and [this](https://github.com/dotnet/corefxlab/pull/2458/files#diff-6f44e9ce636799efd5c8b26382f2248bR42)). That hides a lot of the +1 and -1 making the code easier to read, and they will certainly be inlined.

TylerBrinkley replied:
It doesn't seem worth it as it would only be used once here as the other place I do this I use a ref variable.

TylerBrinkley commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,514 @@
+                TValue value = GetAt(index, out TKey key);
+                return new KeyValuePair<TKey, TValue>(key, value);
+            }
+            set => throw new NotSupportedException(); // Should this be supported?
```
Thoughts on this?

danmosemsft replied:
I think it's fine to throw for now, but it should be InvalidOperationException.

danmosemsft replied:
Scratch that, NotSupportedException is probably better.

TylerBrinkley replied:
I ended up implementing this.

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -399,12 +403,13 @@ private bool TryInsert(int? index, TKey key, TValue value, InsertionBehavior beh
                 }
             }
 
+            ++_version;
```
nit: ++ afterwards.

safern commented on tests/Microsoft.Experimental.Collections.Tests/Microsoft/Collections/Extensions/OrderedDictionaryTests.cs:
```diff
@@ -962,5 +963,102 @@ public void TestLargeOrderedDictionaryOperations()
 }
+
+[DebuggerDisplay("{Key}")]
+internal struct Collider : IEquatable<Collider>, IComparable<Collider>
```
Any reason why this is not part of the tests namespace?

TylerBrinkley replied:
I just mimicked what @danmosemsft did in the DictionarySlim PR. I can pull it into the namespace or perhaps as a nested type.

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,544 @@
+                    if (e.Next == i)
+                    {
+                        e.Next = entry.Next;
+                        j = -1;
```
Should we just call break here? I think it would be clearer than making it -1 and then saying, ahh if it is -1 it will exit the loop.

danmosemsft replied:
maybe, just return.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -109,13 +116,13 @@ public OrderedDictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection, IEq
+                Array.Clear(_entries, 0, count);
                 Count = 0;
             }
-            ++_version;
```
Depends what invalid means, if it means "will fail or produce incorrect results" such as in the case where an insert moves an element along in the array, causing the enumerator to potentially return an element more than once or not at all. In the case of Clear, the MoveNext() checks Count every time `if (_index < _orderedDictionary.Count)` so MoveNext() will return false as if it reached the end, which is arguably correct. If Current is already set, it will not be cleared, but we don't check version in that property getter anyway.

It's probably a bit pedantic, but it seems across our collections we never had a solid definition. I think generally we are trying to relax it a bit (eg we now allow Remove during enumeration on Dictionary)

My care level is low though so either is fine.

safern replied:
Fair enough.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,544 @@
+            }
+            Array.Copy(entries, actualIndex, entries, actualIndex + 1, count - actualIndex);
+
+            ref int bucket = ref _buckets[hashCode % _buckets.Length];
```
Per the perf measurements in DictionarySlim, it might be faster to just store `hashCode % _buckets.Length` and index into `_buckets` again when you want to set it. Perhaps you already tried though.

danmosemsft replied:
Same comment other places. Only measurements can say.

TylerBrinkley replied:
I'll leave it as is until performance tests are added.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,544 @@
+            return true;
+        }
+
+        private void RemoveEntryFromBucket(int i)
```
`i` -> `entryIndex` perhaps?

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,544 @@
+            ref int b = ref _buckets[entry.HashCode % _buckets.Length];
+            if (b == i + 1)
+            {
+                b = entry.Next + 1;
```
`// Bucket was pointing to removed entry. Update it to point to the next in the chain` ?

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,544 @@
+            }
+            else
+            {
+                int j = b - 1;
```
`// Start at the entry the bucket points to, and walk the chain until we find the entry with the index we want to remove, then fix the chain` ?

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,544 @@
+            ref int b = ref _buckets[entry.HashCode % _buckets.Length];
+            if (b == i + 1)
+            {
+                b += incrementAmount;
```
Simliar comments to above?

Thanks for breaking the method into two, now it's clear to me what it does. Maybe there is a way to write it in one method in a clear enough way, but this is fine for now.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/HashHelpers.cs:
```diff
@@ -0,0 +1,93 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
```
Nit, extra newline

TylerBrinkley commented:
I've added a `Move` and `MoveRange` method as proposed in [corefx#14338](https://github.com/dotnet/corefx/issues/14338).

I've also added support for the IList<KeyValuePair<TKey, TValue>>.this[int index] setter. The logic is as follows.

If the key is not contained in the dictionary then the entry at the specified index is replaced.
If the key is contained within the dictionary at the specified index then the key and value are updated for the entry.
If the key is contained within the dictionary at a different index then an exception is thrown as the method shouldn't affect the indices of other entries.

I've also addressed the PR feedback.

omariom commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,678 @@
+                throw new ArgumentOutOfRangeException(nameof(index), Strings.ArgumentOutOfRange_Index);
+            }
+
+            Entry entry = _entries[index];
```
It doesn't check if the entry has been removed? Is it a part of the contract?

TylerBrinkley replied:
When an entry is removed all the entries after it are pushed down so that the entries array is always contiguous.

danmosemsft commented:
We don't need perf tests in order to merge this into corefxlab. But if you're making changes to tune perf, you might want to have them first, since sometimes the effect of a microoptimization can be counterintuitive. They are easy to create and run and have been very helpful in the DictionarySlim PR.

TylerBrinkley commented:
Thanks. I'm not making any changes to tune performance for now, just simplifying the code.

TylerBrinkley commented:
@dotnet-bot retest

TylerBrinkley commented:
@dotnet-bot test

danmosemsft commented:
@dotnet-bot test this please

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.KeyCollection.cs:
```diff
@@ -0,0 +1,135 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Please update header to match
```
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
```

Some of them were wrong a little while ago, maybe you snapped then.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.ValueCollection.cs:
```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
SAme license

danmosemsft commented on tests/Microsoft.Experimental.Collections.Tests/Microsoft/Collections/Extensions/OrderedDictionaryTests.cs:
```diff
@@ -0,0 +1,1492 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
License

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,703 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
License

danmosemsft commented:
Good enough to put out for feedback. @TylerBrinkley any more changes planned right now? Mind if I squash?

danmosemsft commented:
I rebased but it looks like I don't have write permissions to your branch. Please grab it yourself from here: https://github.com/danmosemsft/corefxlab/tree/ordered-dictionary and satisfy yourself I did it correctly 😃

TylerBrinkley commented:
Looks good to me. How do I apply the squash to my branch? A git merge?

YohDeadfall commented:
You can do it via `git rebase -i`, so you would be able to choose what commits to squash.

danmosemsft commented:
I am not a git expert but I would expect you would just add my fork as a remote and fetch it, then reset your branch to it and force push ie
// add my remote and fetch 
git checkout yourbranch
git reset --hard mybranch
git push --force

Does that work? Should be no rebase -I as I did that part.

danmosemsft commented:
BTW would be good to go through and add XML comments to the public members with a note on their complexity?

YohDeadfall commented:
@danmosemsft Sorry, haven't read that you squashed commits already. Then yes, this will do the trick.

TylerBrinkley commented:
Alright, I've grabbed your changes and force pushed them into my branch, makes me think of Star Wars by the way. Now for the XML documentation, do you know where can I find the XML documentation for `Dictionary` so I don't have to recreate the wheel?

danmosemsft commented:
for core, the sources are at 
https://source.dot.net/#System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs,d3599058f8d79be0
or in the corefx repo.

for desktop, either 
https://source.dot.net/#System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs
or
C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETCore\v4.5\System.Collections.xml

TylerBrinkley commented:
Why is there no XML documentation in the source?

danmosemsft commented:
In CoreFX, XML docs are used inconsistently. In this case it is because the code started in Desktop, which was inconsistent because it was not used directly for docs or intellisense at that time. There is a PR up in CoreFX to at least validate what there is. It would be valuable to have an XML docs (at least `<summary>` on anything non trivial) on all public/protected members.

Meanwhile I will find out more about how they generate docs.microsoft.com from the sources.

TylerBrinkley commented:
Thanks, I'll use the System.Collections.xml file for now.

TylerBrinkley commented:
I've added XML Documentation throughout.

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.KeyCollection.cs:
```diff
@@ -11,14 +11,27 @@ namespace Microsoft.Collections.Extensions
+            /// <summary>
+            /// Gets the number of elements contained in the <see cref="OrderedDictionary{TKey, TValue}.KeyCollection" />.
+            /// </summary>
+            /// <returns>The number of elements contained in the <see cref="OrderedDictionary{TKey, TValue}.KeyCollection" />. Retrieving the value of this property is an O(1) operation.</returns>
```
Should `Retrieving the value of this property is an O(1) operation.` be part of the summary tag?

TylerBrinkley replied:
That's how it was specified in the Dictionary XML docs for its KeyCollection. I agree it would make more sense there.

safern commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -214,6 +349,18 @@ public void Insert(int index, TKey key, TValue value)
+        /// -or-
+        /// <paramref name="toIndex"/> is less than 0.
+        /// -or-
+        /// <paramref name="toIndex"/> is equal to or greater than <see cref="OrderedDictionary{TKey, TValue}.Count" /></exception>
```
Can </exception> be in a closing tag for this guys with multiple or statements? It seems more readable.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,981 @@
+        /// </summary>
+        public void Clear()
+        {
+            if (_count > 0)
```
BTW, in DictionarySlim I made Clear() reset the capacity also. Dictionary retains the arrays as you have done. I guess it depends on the use case, whether you are likely to need that capacity soon enough again.

BTW did you mean to increment version?

danmosemsft replied:
I was going to fix it - but your branch is still read only 😱 .  for future reference you can check this box if you want to accept changes to a PR:
![image](https://user-images.githubusercontent.com/6385855/48750953-c3132100-ec36-11e8-88b4-76008dcdbc19.png)

TylerBrinkley replied:
That's weird because the checkbox is checked for me.

danmosemsft commented on src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs:
```diff
@@ -0,0 +1,981 @@
+        /// Returns the zero-based index of the element with the specified key within the <see cref="OrderedDictionary{TKey, TValue}" /> as an O(1) operation.
+        /// </summary>
+        /// <param name="key">The key of the element to locate.</param>
+        /// <returns>The zero-based index of the element with the specified key within the <see cref="OrderedDictionary{TKey, TValue}" />, if found; otherwise, –1.</returns>
```
`-1` on this line is actually `U+2013	EN DASH` followed by `1`. Can you please use regular ASCII hyphen or else the C# `\u2013`? We use `\u` for anything over 0x7F.

TylerBrinkley replied:
Good catch. That was an issue in the source I copied from the XML Documentation on Dictionary.

danmosemsft commented:
@ahsonkhan random termination while running JSON tests - I believe I have seen this a few times before.
https://ci.dot.net/job/dotnet_corefxlab/job/master/job/ubuntu16.04_release_prtest/2824/consoleText

```
Skipped  System.Text.JsonLab.Tests.ReadFromPipeTests.ReadFromPipeUsingSpan
The active test run was aborted. Reason: 

Total tests: Unknown. Passed: 261. Failed: 0. Skipped: 3.
```
Bit fishy

danmosemsft commented:
@dotnet-bot test this please

danmosemsft commented:
nits above, then I will hit merge!

danmosemsft commented:
Thanks @TylerBrinkley ! Once https://github.com/dotnet/corefxlab/pull/2594 is merged then up to NuGet.
